_adr/81/index.adoc

```diff
@@ -19,10 +19,243 @@ NOTE: https://datatracker.ietf.org/doc/html/rfc2119[RFC-2119] keywords are used
+
+For example, for the "Example" series, we might have the sizes:
+
+`small`:: Ingress 30MB/s, egress 30MB/s, 1000 partitions, 500 connections per instance, max 100 connections per second, 1MB max message size.
+`medium`:: Ingress 60MB/s, egress 60MB/s, 2000 partitions, 1000 connections per instance, max 100 connections per second, 1MB max message size.
+`large`:: Ingress 90MB/s, egress 90MB/s, 3000 partitions, 1500 connections per instance, max 100 connections per second, 1MB max message size.
+
+In this example the functional attributes scales linearly, with the exception of connection rate and max message size.
+In general, linearity SHALL NOT be a requirement, even for those attributes which do scale up.
+For example, it would have been acceptable to define the `large` ingress as 100M/s, even though this is not 3× the `small` ingress, even when the other `large` functional parameters are 3× their `small` values.
+
+=== General principles
+
+* A fleet shard is required to provide only a subset of all series supported by the service globally. Indeed, a fleet share might possibly support only a single series one. For example a shard on AWS might support a different set of series to a shard on Azure, or the same series might have different sizes on those clouds.
+
+* The availability of a series might be limited to a particular cloud, region or combination of both.
+
+* It is a private concern of the shard exactly how each size in the series is provisioned. In principle it could use vertical, horizontal scaling and/or density > 1, in addition to applying quotas to limit capacity to the advertised limits. The fleet manager doesn't know.
+
+=== UI/Fleet manager interface
+
+This ADR doesn't mandate any particular presentation for how the scale options are presented in the UI.
+In particular while there is only one series defined, there is no requirement to display series in the UI at all.
+
+The fleet manager SHALL be the source of truth for the series and sizes available, and how those depend on cloud and region. 
+The fleet manager SHALL provide an endpoint which the MUST UI use to determine the available series and sizes.
+The UI MUST NOT make assumptions about linearity of scaling, instead it uses only the series and sizes provided by the endpoint.
+
+==== Series discovery
+
+Reusing the "Example" series defined above, let's illustrate how the Fleet Manager could communicate the available kinds of cluster to a UI.
+
+.Sequence diagram of UI/Fleet manager interaction
+image::ui-fleet_manager.png[UI/Fleet manager interaction]
+
+After the UI has discovered the available cloud providers and regions it would make another GET request `/api/kafkas_mgmt/v1/series/{cloudId}/{cloudRegion}` to discover the series available in this cloud and region.
+An example response JSON is shown below.
+
+[source,json]
+.Example series available to a UI from the Fleet Manager's `GET /api/kafkas_mgmt/v1/series/{cloudId}/{cloudRegion}` endpoint
+----
```

Comment:
I wasn't expecting the sizes to be per cloud provider/region, I thought they were consistent across all cloud providers/regions from what @bibryam said.

Reply:
There's nothing to stop us serving the same set of series from `.../aws/us-east-1` as from some other cloud and region. My thinking for parameterising the endpoint on provider and region was this would allow a future flavour of the service which ran on e.g. ARM silicon and local SSD from those AWS regions which supported it, and for the UI to make that apparent. 

The main downside seems to be the need for an extra request by the UI.

Reply:
_My question is not related to @pmuir's comment above_  

I'm not familiar with the response we get with what's described here:
> After the UI has discovered the available cloud providers and regions it would make another GET request `/api/kafkas_mgmt/v1/series/{cloudId}/{cloudRegion}` to discover the series available in this cloud and region.

Does it make sense to include proposed updates to what would be returned with the list of available cloud providers and regions? Are we expecting any updates with that API call to provide details about available capacity in a region? 

Without seeing both responses (list of cloud providers/regions and list of sizes), it isn't clear to me if we're getting the information we need to communicate to the user something like “this configuration is temporarily unavailable, try a different configuration” before they submit the request.

Reply:
@jgiardino do we do that right now? I know we were thinking about that kind of problem with the dynamic OSD sizing work, but that's on hold at the moment (we'll come back to it soon enough once we have support for larger instances).

I guess we _could_ use this endpoint to express which profiles and sizes were ready for immediate deployment. One problem is just how dynamic that information could be. There will always be the case where it turns out we can't provision an instance when handling an instance creation request. From an API PoV I think it might be nicer to support a `validateOnly` parameter on the creation endpoint, which the UI could poll regularly when the instance creation modal was displayed.

Reply:
We are supposed to be getting information about whether a cloud provider region is available. I remember seeing some discussions on how to update that to support size selection, but I don't remember where.

Maybe @pb82 has more details about the API that provides details about cloud provider regions and how we would use that API along with the API for available sizes to communicate to users what their options are.

Reply:
This is a good point. I'd argue that availability for creating instances should be part of this new endpoint, since for a given cloud+region we might have capacity to create a 1SU/Small instance but not to create a 5SU/3Xlarge instance.

Reply:
is availability a _public_ concept (anyting the UI consumes has to be), or a private concept?

Reply:
Just to be clear, there two things going on here. 

First there is the capability problem: Whether a given cloud + region can support a given size in principal. I.e. we've decided to enable that size for that cloud and region, which is a business decision. We want the ability to limit at least _profiles_ to a cloud and region. e.g. when we roll out support for some enterprisey profile we probably don't want to do that on Azure and AWS _simultaneously_. Or if we ever had a profile that was tied to AWS infra that wasn't available in all regions that we operated then we'd need to be able to express that too.

Then there is the capacity problem: whether we can provision an instance of a certain size _right now_.

I believe both must (eventually) be a public concepts. The UI must be able to inform the user what is possible and what is possible right now. 

The contract being outlined here is about expressing the capability part. We will eventually need to deal with the capacity part. That could be do using the `validateOnly` parameter on the creation endpoint I mentioned, but TBH I see it as out of scope for this ADR.